core/shared/src/main/scala/zio/stm/TQueue.scala

```diff
@@ -18,77 +18,83 @@ package zio.stm
 
 import scala.collection.immutable.{ Queue => ScalaQueue }
 
-import com.github.ghik.silencer.silent
-
 final class TQueue[A] private (val capacity: Int, ref: TRef[ScalaQueue[A]]) {
-  def offer(a: A): STM[Nothing, Unit] =
-    for {
-      q <- ref.get
-      _ <- STM.check(q.length < capacity)
-      _ <- ref.update(_.enqueue(a))
-    } yield ()
-
-  // TODO: Scala doesn't allow Iterable???
-  @silent("enqueueAll")
-  def offerAll(as: List[A]): STM[Nothing, Unit] =
-    ref.update(_.enqueue(as)).unit
-
-  def poll: STM[Nothing, Option[A]] = takeUpTo(1).map(_.headOption)
-
-  def size: STM[Nothing, Int] = ref.get.map(_.length)
 
-  def take: STM[Nothing, A] =
-    ref.get.flatMap { q =>
-      q.dequeueOption match {
-        case Some((a, as)) =>
-          ref.set(as) *> STM.succeed(a)
-        case _ => STM.retry
-      }
-    }
-
-  def takeAll: STM[Nothing, List[A]] =
-    ref.modify(q => (q.toList, ScalaQueue.empty[A]))
+  /**
+   * Checks if the queue is empty.
+   */
+  def isEmpty: STM[Nothing, Boolean] =
+    ref.get.map(_.isEmpty)
 
-  def takeUpTo(max: Int): STM[Nothing, List[A]] =
-    ref.get
-      .map(_.splitAt(max))
-      .flatMap(split => ref.set(split._2) *> STM.succeed(split._1))
-      .map(_.toList)
+  /**
+   * Checks if the queue is at capacity.
+   */
+  def isFull: STM[Nothing, Boolean] =
+    ref.get.map(_.size == capacity)
 
   /**
-   * View the next element in the queue without removing it. retries if the queue is empty.
+   * Views the last element inserted into the queue, retrying if the queue is
+   * empty.
    */
-  def peek: STM[Nothing, A] =
+  def last: STM[Nothing, A] =
     ref.get.flatMap(
-      _.headOption match {
+      _.lastOption match {
         case Some(a) => STM.succeed(a)
         case None    => STM.retry
       }
     )
 
   /**
-   * View the last element inserted into the queue. retries if the queue is empty
+   * Offers the specified value to the queue, retrying if the queue is at
+   * capacity.
    */
-  def last: STM[Nothing, A] =
+  def offer(a: A): STM[Nothing, Unit] =
+    ref.get.flatMap { q =>
+      if (q.length < capacity) ref.update(_.enqueue(a)).unit
+      else STM.retry
+    }
+
+  /**
+   * Offers each of the elements in the specified collection to the queue,
+   * retrying if there is not capacity in the queue for all of the elements in
+   * the collection.
+   */
+  def offerAll(as: Iterable[A]): STM[Nothing, Unit] =
+    ref.get.flatMap { q =>
+      if (as.size <= capacity - q.length) ref.update(_.enqueue(as.toList)).unit
+      else STM.retry
```

Comment:
The other alternative would be placing as many as we can now, and then offeringAll the rest. I think this would result in less surprising behavior (imaging putting 101 elements into a queue with capacity 100—it would never succeed unless you put some of them in and wait to put the rest in).

Reply:
I thought about the same thing but struggled with how to do it in a transactional context. Here is an implementation that tries to do what you describe and a test for it:

```scala
def offerAll(as: Iterable[A]): STM[Nothing, Unit] =
  if (as.isEmpty)
    STM.unit
  else
    ref.get.flatMap { q =>
      val (forQueue, remaining) = as.splitAt(capacity - q.size)
      if (forQueue.isEmpty) offerAll(remaining)
      else ref.set(q.enqueue(forQueue.toList)) *> offerAll(remaining)
    }

testM("offerAll respects capacity") {
  for {
    tq     <- TQueue.bounded[Int](3).commit
    _      <- tq.offerAll(List(1, 2, 3)).commit.fork
    first  <- STM.collectAll(List(tq.take, tq.take, tq.take)).commit
    second <- STM.collectAll(List(tq.take, tq.take)).commit
  } yield assert(first)(equalTo(List(1, 2, 3))) &&
    assert(second)(equalTo(List(4, 5)))
},
```

The challenge is what we do if the queue is at capacity. In the current version, we immediately try to offer again, which results in us just spinning in a tight loop. If instead we retry at that point we risk not being productive, because if we try to put 101 elements into a queue with capacity 100 we're always going to hit a point where the queue doesn't have capacity at which point we will have to retry and start the whole thing over again.

Conceptually it seems like we need a way to suspend until one of the transactional variables has changed without retrying the transaction. Is there a way to do that or is there something I'm missing here?

Reply:
Ah right, the retry will rollback and wait... Possibly we could "fix" it like this:

```scala
def offerAll(as: Iterable[A]): STM[Nothing, Iterable[A]]
```

Then the user can decide what to do with the excess.

Though possibly `offerMax` is a better name for such a thing.

Perhaps for `offerAll`, we can just `STM.die` if the user tries to do something impossible, like offering 10 in a 9 element queue?

Reply:
Yes, done in the latest version. Is it possible to implement the "suspend for a change but don't retry" functionality?

Reply:
Not in STM. That would violate transactionality. Of course we could just return IO for this method. Is that better than dying?

Reply:
I think people are going to want to compose multiple offers and are just going to do it with something like `collectAll` if we don't provide something which will go back to hanging if the number of elements is greater than the capacity. Maybe we want to just provide `offerMax` and maybe rename it to `offerAll`?

Reply:
Ok, sounds good!

Reply:
All set.